planner/core/rule_partition_processor.go

```diff
@@ -351,3 +362,445 @@ func (s *partitionProcessor) findByName(partitionNames []model.CIStr, partitionN
 func (*partitionProcessor) name() string {
 	return "partition_processor"
 }
+
+type lessThanData struct {
+	data     []int64
+	maxvalue bool
+}
+
+func (lt *lessThanData) length() int {
+	return len(lt.data)
+}
+
+func (lt *lessThanData) compare(ith int, v int64) int {
+	if ith == len(lt.data)-1 {
+		if lt.maxvalue {
+			return 1
+		}
+	}
+	switch {
+	case lt.data[ith] > v:
+		return 1
+	case lt.data[ith] == v:
+		return 0
+	}
+	return -1
+}
+
+// partitionRange represents [start, range)
+type partitionRange struct {
+	start int
+	end   int
+}
+
+// partitionRangeOR represents OR(range1, range2, ...)
+type partitionRangeOR []partitionRange
+
+func fullRange(end int) partitionRangeOR {
+	var reduceAllocation [3]partitionRange
+	reduceAllocation[0] = partitionRange{0, end}
+	return partitionRangeOR(reduceAllocation[:1])
+}
+
+func (or partitionRangeOR) intersectionRange(start, end int) partitionRangeOR {
+	// Let M = intersection, U = union, then
+	// a M (b U c) == (a M b) U (a M c)
+	ret := or[:0]
+	for _, r1 := range or {
```

Comment:
just a question: do we can make sure `or` be sorted?

if so maybe we can more optimize to avoid `{6, 7} U {3, 9}` call in `{0, 4}, {6, 7}, {8, 11}  U {3, 9}` or `{3, +INF} U {6, 7}` call in `{3, +INF} U {0, 4}, {6, 7}`- -?

it seems leaf exp (`a > 3` in `a > 3 and b < 5`) can take benfit from this(although `intersectionRange` is fast)

Reply:
Maintain the `sorted` condition is more restrictive.
The `or` array is usually small, so I guess loop over the whole array VS  maintain the `sorted` array and break in advance don't have a big difference.